File: gateway/src/main/kotlin/com/qianlei/gateway/router/Router.kt
package com.qianlei.gateway.router

import com.qianlei.gateway.config.PluginConfig
import com.qianlei.gateway.config.Service
import com.qianlei.gateway.constant.HttpMethod

/**
 *
 * @author qianlei
 */
data class Router(
    val name: String,
    val labels: Map<String, String> = mapOf(),
    val version: String? = null,
    val description: String? = null,
    val publish: Boolean = false,
    val host: String? = null,
    val path: String = "/*",
    val methods: Set<HttpMethod> = emptySet(),
    val service: Service = Service(),
    val plugins: List<PluginConfig> = listOf()
)

File: gateway/router/router.go
package router

import (
	"gateway/pipeline"
	"gateway/service"
)

type Router struct {
	Name        string
	Labels      map[string]string
	Version     string
	Description string
	Publish     bool
	Host        string
	Path        string
	Methods     map[string]bool
	Service     service.Service
	Plugins     []pipeline.Plugin
	pipelines   []pipeline.Pipeline
}

func (r Router) Pipelines() []pipeline.Pipeline {
	if r.pipelines == nil {
		var ans []pipeline.Pipeline
		for i := 0; i < len(r.Plugins); i++ {
			ans = append(ans, r.Plugins[i].Pipelines())
		}
		r.pipelines = ans
	}
	return r.pipelines
}
